### DISCUSSION TITLE
Stripping unicode null character from the TextNode

### DISCUSSION THREAD

**Discussion by pwspot** [FasterXML/jackson-databind#3960] [Q&A]
I am looking for the proper way to strip unicode null (`\u0000`) characters from `TextNode` objects, within the `ObjectNode` (extending `JsonNode`).

Firstly, a JSON String containing `\\u0000` (double backslash is intentional here) is converted into a `JsonNode`. The result object contains a `LinkedHashMap` with an element with a key of type String and a value of type `TextNode`. The `TextNode` unescapes the String and it actually contains a single unicode null character.

Secondly, the resulting object can become a property of another object or an element of a collection. Then this "another object" or a collection is serialized using `valueToTree`. Here, for simplicity, I am skipping this "wrapping" part and it's just calling `valueToTree` over the `JsonNode`:

```
String jsonWithUnicodeNull = "{ \"foo\": \"\\u0000\" }";
ObjectMapper mapper = new ObjectMapper();

// 1
JsonNode node = mapper.readTree(jsonWithUnicodeNull);

// 2
ObjectNode out = mapper.valueToTree(node);
```

<img width="359" alt="image" src="https://github.com/FasterXML/jackson-databind/assets/106061246/6d44cbd8-ebb4-404a-8665-9c134e8262d7">

What I am trying to achieve is to strip the unicode null characters when calling the `valueToTree` over the 2nd object.
I could not find any way to apply a `JsonSerializer` to work with the `TextNode` objects.
I was able to do that by overriding the mapper's `JsonNodeFactory`:
```
mapper.setNodeFactory(new JsonNodeFactory(false) {
  @Override
  public TextNode textNode(String text) {
    return new TextNode(org.apache.commons.lang3.StringUtils.remove(text, '\u0000'));
  }
});
```
What I don't like with this approach is the assumption on what the original node factory object is and its defaults.

For the context, these are another settings that I use with the mapper object:
```
mapper.configure(SerializationFeature.EAGER_SERIALIZER_FETCH, true); // is default
mapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
mapper.setSerializationInclusion(Include.NON_NULL);
mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```

I am not stripping the character in the first step (`readTree`) as this actually can come from multiple sources and filtering it in a single place seems more reasonable.

### Version information
Which Jackson version(s) was this for? 2.14.2
JDK - Temurin-11.0.15+10

**Comment by cowtowncoder** [comment]
My suggestion would be read content as `JsonNode` and then post-process it by traversing, to change `TextNode`s where necessary.

**Reply by pwspot** [reply]
Could you guide me to any example of post-processing?

**Reply by cowtowncoder** [reply]
Just traverse over values recursively (need to handle `ArrayNode` and `ObjectNode` specifically), look for `TextNode`s, replace if change needed.
